presto-hive/src/test/java/io/prestosql/plugin/hive/util/TestThrottledAsyncQueue.java

```diff
@@ -0,0 +1,266 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.prestosql.plugin.hive.util;
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.util.concurrent.ListenableFuture;
+import io.airlift.concurrent.Threads;
+import org.testng.annotations.AfterClass;
+import org.testng.annotations.BeforeClass;
+import org.testng.annotations.Test;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+
+import static io.airlift.concurrent.MoreFutures.getFutureValue;
+import static io.airlift.testing.Assertions.assertContains;
+import static org.testng.Assert.assertEquals;
+import static org.testng.Assert.assertFalse;
+import static org.testng.Assert.assertTrue;
+import static org.testng.Assert.fail;
+
+public class TestThrottledAsyncQueue
+{
+    private ExecutorService executor;
+
+    @BeforeClass
+    public void setUpClass()
+    {
+        executor = Executors.newFixedThreadPool(8, Threads.daemonThreadsNamed("test-async-queue-%s"));
+    }
+
+    @AfterClass(alwaysRun = true)
+    public void tearDownClass()
+    {
+        executor.shutdownNow();
+    }
+
+    @Test(timeOut = 10_000)
+    public void testThrottle()
+    {
+        // Make sure that the dequeuing is throttled even if we have enough elements in the queue
+
+        ThrottledAsyncQueue<Integer> queue = new ThrottledAsyncQueue<>(3, 10, executor);
+        assertTrue(queue.offer(1).isDone());
+        assertTrue(queue.offer(2).isDone());
+        assertTrue(queue.offer(3).isDone());
+        assertTrue(queue.offer(4).isDone());
+        assertTrue(queue.offer(5).isDone());
+        assertTrue(queue.offer(6).isDone());
+        queue.finish();
+
+        // no throttling, enough elements in the queue
+        ListenableFuture<List<Integer>> future1 = queue.getBatchAsync(2);
+        assertTrue(future1.isDone());
+        assertEquals(getFutureValue(future1), ImmutableList.of(1, 2));
+        assertFalse(queue.isFinished());
+
+        // we can only dequeue one more element before being throttled
+        ListenableFuture<List<Integer>> future2 = queue.getBatchAsync(2);
+        assertFalse(future2.isDone());
+        assertEquals(getFutureValue(future2), ImmutableList.of(3, 4));
+        assertFalse(queue.isFinished());
+
+        // we are now throttled, this future will not be immediate
+        ListenableFuture<List<Integer>> future3 = queue.getBatchAsync(2);
+        assertFalse(future3.isDone());
+        assertEquals(getFutureValue(future3), ImmutableList.of(5, 6));
+        assertTrue(queue.isFinished());
+    }
+
+    @Test(timeOut = 10_000)
+    public void testThrottleEmptyQueue()
+            throws Exception
+    {
+        // Make sure that the de-queuing is throttled even if we have enough elements in the queue
```

Comment:
Should this be `even if we do not have enough`?

Reply:
done, I improved the comment